container/memory_store.go

```diff
@@ -1,67 +1,78 @@
 package container
 
-import "sync"
+import (
+	"sync"
+
+	"github.com/docker/docker/pkg/locker"
+)
 
 // memoryStore implements a Store in memory.
 type memoryStore struct {
-	s map[string]*Container
-	sync.Mutex
+	s          map[string]*Container
+	globalLock sync.Mutex
+	namedLock  *locker.Locker
 }
 
 // NewMemoryStore initializes a new memory store.
 func NewMemoryStore() Store {
 	return &memoryStore{
-		s: make(map[string]*Container),
+		s:         make(map[string]*Container),
+		namedLock: locker.New(),
 	}
 }
 
 // Add appends a new container to the memory store.
 // It overrides the id if it existed before.
 func (c *memoryStore) Add(id string, cont *Container) {
-	c.Lock()
+	c.globalLock.Lock()
 	c.s[id] = cont
-	c.Unlock()
+	c.globalLock.Unlock()
 }
 
 // Get returns a container from the store by id.
 func (c *memoryStore) Get(id string) *Container {
-	c.Lock()
+	c.namedLock.Lock(id)
+	c.globalLock.Lock()
 	res := c.s[id]
-	c.Unlock()
+	c.globalLock.Unlock()
+	c.namedLock.Unlock(id)
 	return res
 }
 
 // Delete removes a container from the store by id.
 func (c *memoryStore) Delete(id string) {
-	c.Lock()
+	c.namedLock.Lock(id)
+	c.globalLock.Lock()
 	delete(c.s, id)
-	c.Unlock()
+	c.globalLock.Unlock()
+	c.namedLock.Unlock(id)
 }
 
 // List returns a sorted list of containers from the store.
 // The containers are ordered by creation date.
 func (c *memoryStore) List() []*Container {
 	containers := new(History)
-	c.Lock()
+	c.globalLock.Lock()
 	for _, cont := range c.s {
 		containers.Add(cont)
 	}
-	c.Unlock()
+	c.globalLock.Unlock()
 	containers.sort()
 	return *containers
 }
 
 // Size returns the number of containers in the store.
 func (c *memoryStore) Size() int {
-	c.Lock()
-	defer c.Unlock()
-	return len(c.s)
+	c.globalLock.Lock()
+	l := len(c.s)
+	c.globalLock.Unlock()
```

Comment:
Yea! I saw that this morning. Might as well link it again http://lk4d4.darth.io/posts/defer/